go/vt/topotools/vschema_ddl.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2018 The Vitess Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topotools
+
+import (
+	"reflect"
+
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+
+	vschemapb "vitess.io/vitess/go/vt/proto/vschema"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+// ApplyVSchemaDDL applies the given DDL statement to the vschema
+// keyspace definition and returns the modified keyspace object.
+func ApplyVSchemaDDL(ksName string, ks *vschemapb.Keyspace, ddl *sqlparser.DDL) (*vschemapb.Keyspace, error) {
+	if ks == nil {
+		ks = new(vschemapb.Keyspace)
+	}
+
+	if ks.Tables == nil {
+		ks.Tables = map[string]*vschemapb.Table{}
+	}
+
+	if ks.Vindexes == nil {
+		ks.Vindexes = map[string]*vschemapb.Vindex{}
+	}
+
+	var tableName string
+	var table *vschemapb.Table
+	if !ddl.Table.IsEmpty() {
+		tableName = ddl.Table.Name.String()
+		table, _ = ks.Tables[tableName]
+	}
+
+	switch ddl.Action {
+	case sqlparser.CreateVindexStr:
+		name := ddl.VindexSpec.Name.String()
+		if _, ok := ks.Vindexes[name]; ok {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex %s already exists in keyspace %s", name, ksName)
+		}
+
+		// Make sure the keyspace has the sharded bit set to true
+		// if this is the first vindex defined in the keyspace.
+		if len(ks.Vindexes) == 0 {
+			ks.Sharded = true
```

Comment:
I was thinking that it would make sense this to be driven by a flag. That way we can be explicit with the intent. If the flag is not provided and the Vindexes len is 0, adding a vindex should fail.

Reply:
By definition, vindexes are only present in Sharded keyspaces, so I think the act of creating a vindex expresses the intent of whether we want

I have ideas about how we can extend the vschema ddl syntax to support unsharded keyspaces, but even then we won't have vindexes and I think above will hold.